File: dailyQuestion/2020-09/9-01/python/solution_recurse.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def pathSum(root: TreeNode, sum: int) -> [[int]]:
    temp = []
    result = []
    def dfs(root: TreeNode, left: int) :
        if not root :
            return
        left -= root.val
        temp.append(root.val)
        if not root.left and not root.right and left == 0 :
            result.append(temp[:])
        dfs(root.left,left)
        dfs(root.right,left)
        temp.pop()

    dfs(root,sum)
    return result

if __name__ == "__main__" :
    node = TreeNode(3)
    node.left = TreeNode(5)
    node.right = TreeNode(1)
    node.left.left = TreeNode(6)
    node.left.right = TreeNode(2)
    node.left.right.left = TreeNode(7)
    node.left.right.right = TreeNode(4)

    node.right.left = TreeNode(0)
    node.right.right = TreeNode(8)
    result = pathSum(node,12)
    print(result)

File: dailyQuestion/2020/2020-09/9-01/golang/solution_recurse.go
package main

import "fmt"

func main() {
	root := &TreeNode{Val: 3, Left: &TreeNode{Val: 5}, Right: &TreeNode{Val: 1}}
	root.Left.Left = &TreeNode{Val: 6}
	root.Left.Right = &TreeNode{Val: 2, Left: &TreeNode{Val: 7}, Right: &TreeNode{Val: 4}}
	root.Right.Left = &TreeNode{Val: 0}
	root.Right.Right = &TreeNode{Val: 8}

	result := pathSum(root, 12)
	fmt.Println(result)

}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, sum int) (result [][]int) {
	temp := []int{}
	var dfs func(root *TreeNode, left int)
	dfs = func(root *TreeNode, left int) {
		if root == nil {
			return
		}
		left -= root.Val
		temp = append(temp, root.Val)
		defer func() {
			temp = temp[:len(temp)-1]
		}()
		if root.Left == nil && root.Right == nil && left == 0 {
			result = append(result, append([]int(nil), temp...))
			return
		}
		dfs(root.Left, left)
		dfs(root.Right, left)
	}
	dfs(root, sum)
	return
}
